src/D2L.CodeStyle.TestAnalyzers/ServiceLocator/CustomTestServiceLocatorAnalyzer.cs

```diff
@@ -0,0 +1,152 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using D2L.CodeStyle.TestAnalyzers.Common;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.TestAnalyzers.ServiceLocator {
+
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class CustomTestServiceLocatorAnalyzer : DiagnosticAnalyzer {
+
+		private const string TestServiceLocatorFactoryType
+			= "D2L.LP.Extensibility.Activation.Domain.TestServiceLocatorFactory";
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+			=> ImmutableArray.Create( Diagnostics.CustomServiceLocator );
+
+		private readonly bool _excludeKnownProblems;
+
+		public CustomTestServiceLocatorAnalyzer() : this( true ) { }
+
+		public CustomTestServiceLocatorAnalyzer( bool excludeKnownProblemFixtures ) {
+			_excludeKnownProblems = excludeKnownProblemFixtures;
+		}
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction(
+				RegisterServiceLocatorAnalyzer
+			);
+		}
+
+		public void RegisterServiceLocatorAnalyzer(
+			CompilationStartAnalysisContext context
+		) {
+			INamedTypeSymbol factoryType = context.Compilation
+				.GetTypeByMetadataName( TestServiceLocatorFactoryType );
+
+			if( factoryType == null || factoryType.Kind == SymbolKind.ErrorType ) {
+				return;
+			}
+
+			context.RegisterSyntaxNodeAction(
+				ctx => PreventCustomLocatorUsage(
+					ctx,
+					factoryType
+				),
+				SyntaxKind.InvocationExpression
+			);
+		}
+
+		// Prevent static usage of TestServiceLocator.Create() methods.
+		private void PreventCustomLocatorUsage(
+			SyntaxNodeAnalysisContext context,
+			INamedTypeSymbol disallowedType
+		) {
+			ExpressionSyntax root = context.Node as InvocationExpressionSyntax;
+			if( root == null ) {
+				return;
+			}
+
+			SymbolInfo symbolInfo = context.SemanticModel.GetSymbolInfo( root );
+
+			IMethodSymbol method = symbolInfo.Symbol as IMethodSymbol;
+
+			if( method == null ) {
+				if( symbolInfo.CandidateSymbols == null ) {
+					return;
+				}
+
+				if( symbolInfo.CandidateSymbols.Length != 1 ) {
+					return;
+				}
+
+				// This happens on method groups, such as
+				// Func<IFoo> fooFunc = OldAndBrokenServiceLocator.Get<IFoo>;
+				method = symbolInfo.CandidateSymbols.First() as IMethodSymbol;

```

Comment:
Huh, didn't realize this was how that shook out. So `.Symbol` is actually `null` here?

Reply:
I guess this came from the `OldAndBrokenServiceLocator` analyzer so yup

Reply:
My testing suggested that virtually everything falls under the case where `.Symbol` isn't `null`, but I figured I would play it safe in case there were use cases I had missed. I should edit the comment though...